### PR TITLE
build: configured lavamoat

### LINKED ISSUE #409: Setup LavaMoat
> I accidentally disabled [lavamoat](https://github.com/LavaMoat/LavaMoat) on the core dollar contracts at some point while converting this into a monorepo. We should definitely re-add to the whole project.

_Originally posted by @pavlovcik in https://github.com/ubiquity/ubiquity-dollar/discussions/197#discussioncomment-3641178_

### PR DESCRIPTION
**Pull request by FibrinLab**
Resolves #409

### CODE REVIEW

File: .yarnrc [hunk 1]
```diff
@@ -0,0 +1 @@
+ignore-scripts true
```
Comment (0x4007) [inline .yarnrc:1]: What is this?

File: .yarnrc.yml [hunk 2]
```diff
@@ -25,3 +25,4 @@ plugins:
     spec: "https://yarn.build/latest"
 
 yarnPath: .yarn/releases/yarn-3.2.3.cjs
+enableScripts: false
```
Comment (0x4007) [inline .yarnrc.yml]: What is this?

Comment (FibrinLab) [reply .yarnrc.yml]: It was auto generated by Lavamoat

Comment (FibrinLab) [reply .yarnrc.yml]: Allows only authorized lifecycle hooks to run. All auto generated

**Comment by 0x4007** [comment]
@zgorizzo69 I requested your eyes on this because you set up lavamoat the first time around. Is this currently set up correctly?

**Comment by 0x4007** [comment]
@ FibrinLab its not clear to me how we're supposed to run this. I expect that you would either include a script in package.json or include some type of commit hook.

File: package.json [hunk 3]
```diff
@@ -34,7 +37,9 @@
     "lint:all": "yarn workspaces foreach run lint",
     "build:ci:cloudflare": "bash ./utils/build_cloudflare.sh",
     "build:ci:netlify": "yarn build:all",
-    "postinstall": "husky install"
+    "postinstall": "husky install",
+    "lavamoat-policy": "lavamoat app.js --autopolicy",
+    "start": "lavamoat app.js"
```
Comment (0x4007) [inline package.json]: `Error: ENOENT: no such file or directory`

**Review by 0x4007 — DISMISSED** [review]

**Comment by 0x4007** [comment]
Please re-request a review from me when this is stable.

File: .yarnrc.yml (hunk 2 above)
Comment (Steveantor) [reply .yarnrc.yml]: > Allows only authorized lifecycle hooks to run. All auto generated

you can't simply arbitrarily disable all pre-post scripts.

**Comment by FibrinLab** [comment]
Summary of my PR.

Lavamoat is a set of tools for mitigating supply chain attacks. It does this via two main ways:

ability to control lifecycle scripts. Provided by [@lavamoat/allow-scripts](https://github.com/LavaMoat/LavaMoat/blob/main/packages/allow-scripts)
ability to freeze objects and control permitted privileges. Requires a runtime. Provided by [lavamoat-node](https://github.com/LavaMoat/LavaMoat/blob/main/packages/node)
This PR integrates @lavamoat/allow-scripts and configures it to run only dependency lifecycle hooks that has been specified.

1. The tool works on the assumption that running lifecycle scripts has been disabled. By using --ignore-scripts flag or by configuration via .yarnrc or .npmrc.
2. The module @lavamoat/preinstall-always-fail is to prevent accidentally running yarn | yarn install without disabling lifecycle script. With this module as dependency, the build fails on yarn | yarn install if lifecycle script is not disabled. I proceeded to update the "Install and Build Workflow" to ignore and run whitelisted scripts immediately.
3. Those lifecycle scripts now need to be explicitly triggered by yarn allow-scripts
4. Apologies for accidentally disabling scripts in the .yarnrc.yml file.

Looking forward to comments, corrections and reviews.

**Comment by Steveantor** [comment]
I still don't see allow-scripts for important packages/add-ons, such as those that are built with gyp.

**Comment by FibrinLab** [comment]
> I still don't see allow-scripts for important packages/add-ons, such as those that are built with gyp.

Added packages to the allowScript.

**Comment by Steveantor** [comment]
> > I still don't see allow-scripts for important packages/add-ons, such as those that are built with gyp.
> 
> Added packages to the allowScript.

- `napi-macros-example` might be needed.

- `styled-components` can be safely ignored.

**Comment by FibrinLab** [comment]
> > > I still don't see allow-scripts for important packages/add-ons, such as those that are built with gyp.
> > 
> > 
> > Added packages to the allowScript.
> 
> * `napi-macros-example` might be needed.
> * `styled-components` can be safely ignored.

Okay. ✅ I modified the allow-list further

**Comment by 0x4007** [comment]
Thanks for your helpful feedback @Steveantor. @zgorizzo69 did you get a chance to check out this PR?

**Comment by FibrinLab** [comment]
> Thanks for your helpful feedback @Steveantor. @zgorizzo69 did you get a chance to check out this PR?

Any other corrections?

**Review by zgorizzo69 — APPROVED** [review]
onfiguration looks good as is.
Would be nice to have a [policy]( https://github.com/LavaMoat/LavaMoat/tree/main/packages/node#recommended-usage) but I  don't know how to generate it as we are using next.js

**Comment by FibrinLab** [comment]
> onfiguration looks good as is. Would be nice to have a [policy](https://github.com/LavaMoat/LavaMoat/tree/main/packages/node#recommended-usage) but I don't know how to generate it as we are using next.js

I researched this. I don't think Next js is fully supported yet. Maybe future updates

**Comment by 0x4007** [comment]
Currently unable to use my computer at the moment  

I tried merging but got this error so will try again later. 

@0xcodercrane perhaps some config was changed in your CI?

Be sure to post your wallet address on the issue (not this PR) @FibrinLab 

image

**Comment by Steveantor** [comment]
> Currently unable to use my computer at the moment
> 
> I tried merging but got this error so will try again later.
> 
> @0xcodercrane perhaps some config was changed in your CI?
> 
> Be sure to post your wallet address on the issue (not this PR) @FibrinLab
> 
> image


Since this PR modifies the workflow, your OAuth token doesn't have workflow modification permission.
image

More info: [How to resolve "refusing to allow an OAuth App to create or update workflow" on git push](https://stackoverflow.com/questions/64059610/how-to-resolve-refusing-to-allow-an-oauth-app-to-create-or-update-workflow-on)
@pavlovcik

**Comment by 0x4007** [comment]
Thank you for the helpful information @Steveantor 

I will address this when I'm on my computer next time!

> > Currently unable to use my computer at the moment
> 
> > 
> 
> > I tried merging but got this error so will try again later.
> 
> > 
> 
> > @0xcodercrane perhaps some config was changed in your CI?
> 
> > 
> 
> > Be sure to post your wallet address on the issue (not this PR) @FibrinLab
> 
> > 
> 
> > image
> 
> 
> 
> 
> 
> Since this PR modifies the workflow, your OAuth token doesn't have workflow modification permission.
> 
> image
> 
> 
> 
> More info: [How to resolve "refusing to allow an OAuth App to create or update workflow" on git push](https://stackoverflow.com/questions/64059610/how-to-resolve-refusing-to-allow-an-oauth-app-to-create-or-update-workflow-on)
> 
> @pavlovcik

**Comment by 0x4007** [comment]
@Steveantor I didn't change the permissions but I did the merge from my computer instead of the GitHub iOS app. Let's discuss in https://github.com/ubiquity/ubiquity-dollar/issues/441

**Pull request merged**